flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobVertex.java

```diff
@@ -143,27 +145,27 @@ public JobVertex(String name) {
 	public JobVertex(String name, JobVertexID id) {
 		this.name = name == null ? DEFAULT_NAME : name;
 		this.id = id == null ? new JobVertexID() : id;
-		// the id lists must have the same size
-		this.operatorIDs.add(OperatorID.fromJobVertexID(this.id));
-		this.operatorIdsAlternatives.add(null);
+		OperatorIdPair operatorIdPair = new OperatorIdPair(OperatorID.fromJobVertexID(this.id), null);
+		this.operatorIdPairList = new OperatorIdPairList(Collections.singletonList(operatorIdPair));
 	}
 
 	/**
 	 * Constructs a new job vertex and assigns it with the given name.
 	 *
 	 * @param name The name of the new job vertex.
 	 * @param primaryId The id of the job vertex.
-	 * @param alternativeIds The alternative ids of the job vertex.
 	 * @param operatorIds The ids of all operators contained in this job vertex.
 	 * @param alternativeOperatorIds The alternative ids of all operators contained in this job vertex-
 	 */
-	public JobVertex(String name, JobVertexID primaryId, List<JobVertexID> alternativeIds, List<OperatorID> operatorIds, List<OperatorID> alternativeOperatorIds) {
+	public JobVertex(String name, JobVertexID primaryId, List<OperatorID> operatorIds, List<OperatorID> alternativeOperatorIds) {
 		Preconditions.checkArgument(operatorIds.size() == alternativeOperatorIds.size());
 		this.name = name == null ? DEFAULT_NAME : name;
 		this.id = primaryId == null ? new JobVertexID() : primaryId;
-		this.idAlternatives.addAll(alternativeIds);
-		this.operatorIDs.addAll(operatorIds);
-		this.operatorIdsAlternatives.addAll(alternativeOperatorIds);
+		List<OperatorIdPair> operatorIdPairs = new ArrayList<>();
+		for (int i = 0; i < operatorIds.size(); i++) {
+			operatorIdPairs.add(new OperatorIdPair(operatorIds.get(i), alternativeOperatorIds.get(i)));
+		}
+		this.operatorIdPairList = new OperatorIdPairList(operatorIdPairs);
```

Comment:
Ideally, this should be done by the caller (i.e. `StreamingJobGraphGenerator.createJobVertex`) so the constructor receives a list of pairs.
Also, lists here don't guarantee constant time `get(int)` and it's good to check that both lists have the same size.

Reply:
done